[WIP] Simulate from posterior of the coefficients and smoothing parameters

cbrummitt opened the pull request:
## Summary

This is a work in progress to implement #113. This PR implements the following:

1. a new method for `GAM` called `simulate_from_coef_posterior_conditioned_on_smoothing_parameters` that **simulates from the posterior distribution over the parameters conditioned on the smoothing parameters** using `np.random.multivariate_normal(self.coef_, self.statistics_['cov'], size=n_simulations)`.
2. a new method for every distribution called `sample` that **draws random samples from the given array of expected values**.

## Some questions to consider

1. Right now, `Distribution.sample` does not take a `size` argument. In every `sample` method, the `size` argument is always `None`, so that `numpy` simply broadcasts across the array `mu` and draws one random sample for each `mu`. This is all we need if we just use `distribution.sample(mu)` from the method `GAM.simulate_from_coef_posterior_conditioned_on_smoothing_parameters`: we draw a certain number of random samples of the coefficients, and for each of those samples of the coefficients we draw one sample from the distribution of the response. **Perhaps one may want to add more control over how many samples from the response distribution one would like for each sample from the posterior of the coefficients.**
2. For the `hepatitis_A_bulgaria` dataset with a `LinearGAM`, `simulate_from_coef_posterior_conditioned_on_smoothing_parameters` gives the warning `RuntimeWarning: covariance is not positive-semidefinite
  self.coef_, self.statistics_['cov'], size=n_simulations)`. (This happens no matter whether `constraints` is `'monotonic_inc'`, `'concave'`, or `None`.) The method still returns random samples. The default behavior of `numpy.random.multivariate_normal` is to raise a warning (`check_valid='warn'`). **Is this behavior good enough?** I am not familiar with how severe this problem is.
3. **Should we rename `simulate_from_coef_posterior_conditioned_on_smoothing_parameters`?** It's long but descriptive. It'd be nice to have it called just `simulate_from_posterior`, and if we implement the bootstrap samples to simulate from the distribution over the smoothing parameters, too, then that could become just an optional argument in this catch-all method `simulate_from_posterior`.
4. I made `sample` an abstract method for the `Distribution` class. Feel free to remove that abstract method and the two imports (`from abc import ABCMeta, abstractmethod`) if you don't want to bother to force subclasses of Distribution to implement these methods (because we don't expect new distributions to be written?).
5. `simulate_from_coef_posterior_conditioned_on_smoothing_parameters` checks that the model is already fitted and that `n_simulations` is not `<= 0`. I also copied and pasted the code for `X = check_X(X, ...)`. Is this checking good enough?
6. Should we write tests? We could verify shapes of return values and maybe also verify the random samples are close enough to what we expect.
7. Should we implement bootstrap samples to get some uncertainty over the smoothing parameters? That may be somewhat involved to make a good API, given how computationally expensive it could be, so perhaps that is left for another PR.


## Example
Here's an example of `simulate_from_coef_posterior_conditioned_on_smoothing_parameters` on the first example in the README:

![example](https://user-images.githubusercontent.com/1268375/28783178-76b39a90-75dd-11e7-9768-433332e562dc.png)

The extra code added was:
```
response_simulations = gam.simulate_from_coef_posterior_conditioned_on_smoothing_parameters(
    XX, 100)
for response in response_simulations:
    plt.scatter(XX, response, alpha=.01, color='k')
```
The light-opacity black disks are random samples from the posterior.

I made the first axis of the return value the simulations, so you can loop over simulations with `response in response_simulations`, or compute averages across simulations with `np.mean(response_simulations, axis=0)`, and so on. This convention (that axis 0 is the simulation index) matches that used by [PyMC3's `sample` function](https://pymc-devs.github.io/pymc3/notebooks/getting_started.html#Gradient-based-sampling-methods).

dswah commented on pygam/pygam.py:
```diff
@@ -1781,6 +1781,59 @@ def gridsearch(self, X, y, return_scores=False, keep_best=True,
+            raise ValueError('The number of simulations must be positive; '
+                             'got n_simulations = {}'.format(n_simulations))
+
+        X = check_X(X, n_feats=len(self._n_coeffs) - self._fit_intercept,
```
hell yeah. looks good

dswah commented on pygam/pygam.py:
```diff
@@ -1781,6 +1781,59 @@ def gridsearch(self, X, y, return_scores=False, keep_best=True,
+        if not self._is_fitted:
+            raise AttributeError('GAM has not been fitted. Call fit first.')
+
+        if n_simulations <= 0:
```
👍 
this should be good enough.
I dont feel like checking if its an `int` etc. numpy will check for us.

dswah commented:
@cbrummitt fucking awesome.
I need to figure out the best way to contribute to this PR. Perhaps i need to fork your fork and then branch your branch?

Notes:

1. Nice work on the distribution sample methods!! (not to mention fixing all of those typos!). I see what you're getting at with the size of the sampling and I'm trying to think of the tradeoffs... How could we benefit from sampling more at one node of the bayesian graph than another?
ie why might we want to sample `y_i` multiple times per `mu_i`?
```
mu_i ~ N(mu, cov)
y_i ~ f(mu_i, scale)
```
I don't know the answer or the tradeoffs right now... But I suspect that we may not want to sample multiple `y_i`s per `mu_i` because it introduces correlations into our samples...

Perhaps our `Distribution`s should admit a `size` parameter (for convenience), but we should keep the `size` as `None` for the sake of this method?

2. i think i know why we get those non-positive-definite warnings. i did a quick search and people suggest that it may be because the eigenvalues of the problem are positive but almost zero, which causes numerical problems. i confirmed this for the hepatitis dataset [1: code]:
![image](https://user-images.githubusercontent.com/11619412/28790322-7abea404-75dd-11e7-9a75-f7bd891f92c6.png)

for now we can add a fixed value to the diagonal of the identity matrix. Whenever Wood checks for numerical issues, he looks to see if any eigenvalue is smaller than `sqrt(machine_epsilon)`. On my machine i have `sqrt(machine_epsilon) ~= 1e-8`. so perhaps we should do 
```
n_features = len(self._dtypes) 
cov = self.statistics_['cov'] + np.eye(n_features) * np.sqrt(EPS)
```
(aside, it's looking more and more like it would be good to keep track of the number of features since it comes up quite frequently...)

3. I am a fan of a very simple method name, like `self.sample`. I don't think this will be ambiguous since the API / method signature clearly explains what it's doing, and it's very nice to type :)

4.  im not familiar with ABCs (the documentation is a bit esoteric for me right now 😄), but i'm game. is the benefit that we are guaranteeing that any new `Distribution` will have a `self.sample()` ?(https://docs.python.org/2/library/abc.html#abc.abstractmethod) 
- the automatic testing found an error [2: error] with py2 and the syntax of the ABC. Do you think we need to use the following for python2 compatibility?

5. awesome 👍  i read this and it looks good. i don't think it needs any other checks right now, but i will think about it.

6. tests could be nice. even a simple one like testing the shape would be great, especially when we add the bootstrapping functionality. hmm, if we have some extra time on our hands we might check that the samples are distributed correctly... we could fix the random seed of the test so that we have fewer things to debug in the test, but adding uncertainty to the testing is always a drag. maybe we can punt this second test for later?

7.  i agree, implementing the bootstrapping can wait for a subsequent PR

```
class Distribution(Core):
    __metaclass__ = ABCMeta
```
instead of:

```
class Distribution(Core, metaclass=ABCMeta):
```


[1: code]
```
hep = pd.read_csv('datasets/hepatitis_A_bulgaria.csv').astype(float)

# eliminate 0/0
mask = (hep.total > 0).values
hep = hep[mask]

X = hep.age.values[:,None]
y = hep.hepatitis_A_positive.values / hep.total.values

gam = LinearGAM().fit(X, y)
np.linalg.eigvals(gam.statistics_['cov'])
```

[2: error] 
```
E   ConftestImportFailure: SyntaxError('invalid syntax', ('/home/travis/build/dswah/pyGAM/pygam/distributions.py', 16, 35, 'class Distribution(Core, metaclass=ABCMeta):\n'))
E     File "/home/travis/build/dswah/pyGAM/pygam/__init__.py", line 7, in <module>
E       from pygam.pygam import GAM
E     File "/home/travis/build/dswah/pyGAM/pygam/pygam.py", line 26, in <module>
E       from pygam.distributions import Distribution
```

dswah commented:
as a side note and a reminder to self:

- mvnorn can be sped up, which comes in handy when the covariance matrices are large (i have the code for that, but that's an optimization that should go in a subsequent pull request)

dswah commented:
@cbrummitt is it alright with you if I push code directly to this branch?

it shouldn't be too hard to avoid merge conflicts if our commits are small and atomic, and we commit/pull frequently. 

you've already written the bulk of the code for this PR so we only have a few small things to correct.

let me know. 

I will try and squash the `ABC` bug on python 2.

cbrummitt commented:
Re:
> Perhaps our `Distribution`s should admit a `size` parameter (for convenience), but we should keep the `size` as `None` for the sake of this method?
I looked into the size parameter. It doesn't seem useful after all:
![size_parameter](https://user-images.githubusercontent.com/1268375/28805989-7f127464-763b-11e7-870d-cffc7250f14a.png)
`size=1` and `size=4` also led to `ValueError`s.

So I suggest we punt on exposing the `size` parameter as an optional argument in the `Distribution`s for now and add that later if we see a use case for it. People can always call `sample` more than once (though that's slower than computing a bunch of random samples all at once).

dswah commented:
cool, problem solved ;)

cbrummitt commented:
@dswah Yes, please do feel free to push code into this branch `simulate-from-posterior`. 

I like your idea of renaming the method to `sample`. Then `GAM.sample` calls `Distribution.sample` 🙂  Then later we consider adding bootstrap samples in a backward-compatible way using optional arguments in `sample`? I need to sit down and digest the algorithm on pages 198–199 better. I think we may want to write a rough draft of this bootstrap version before we merge this really simple version, to see whether we can shoe-horn it into the same `sample` method.

dswah commented:
@cbrummitt that's a good idea.
i'm reading the algorithm again, and it's more involved than i believed at first.

dswah commented:
@cbrummitt here is my understanding of the code example on pg 257

His example is custom to a poisson distribution, so we'd need to make sure our bootstrapping makes sense.

It looks like the scheme is to always generate bootstrap samples from the original fitted GAM (ie `self`), then fit a new GAM to that response data. This yields a new vector of smoothing parameters. We then fit another GAM to the original response data with these new smoothing parameters, and sample model coefficients from their conditional distribution and store those samples.
- repeat a few times

at the end we can return all of the sampled coefficients (like a pymc trace), or we can put them through the link function and pass them to the `distribution.sample` method to generate a prediction distribution.

Is that right?

```
def sample_mu(self, X, y, weights=None, n_samples=100, n_bootstraps=20):
    """this is the method from page 257
    """

    if not self._is_fitted:
        raise AttributeError('GAM has not been fitted. Call fit first.')
    
    mu = self.predict_mu(X)
    coefs = []
    for _ in range(n_bootstraps - 1):
        
        # generate bootstrap data
        self.distribution.sample(mu)
        
        # fit smoothing parameters on the bootstrap data
        gam = deepcopy(self)
        gam.set_params(self.get_params())
        gam.gridsearch(X, y_bootstrap, weights=weights)
        lam = gam.lam
        
        # fit coeficients on the original data given the smoothing params
        gam = deepcopy(self)
        gam.set_params(self.get_params())
        gam.lam = lam
        gam.fit(X, y, weights=weights)
        
        # sample from coeficients
        coefs.append(np.random.multivariate_normal(gam.coef_, 
                                                   gam.statistics_['cov'], 
                                                   size=n_samples))
    
    coefs.append(self.sample_from_conditional(X, n_samples=n_samples))
    
    coefs = np.stack(coefs)
    linear_predictor = coefs.dot(X)
    
    return self.link.mu(linear_predictor, self.distribution)
    
def sample(self, X, y, weights=None, n_samples=100, n_bootstraps=20):
    mu = self.sample_mu(X, y, weights=None, n_samples=100, n_bootstraps=20)
    self.distribution.sample(mu, self.scale)
```

cbrummitt commented:
Nice work @dswah! I added the missing `y_bootstrap = ` in the line after the comment `# generate bootstrap data`. I'm editing a local copy of your recent code snippet and will share it here soon (in 1.5 hours I hope).

cbrummitt commented:
`n_samples` is used as a variable for `X.shape[0]`, so I'm thinking it would be clearer to use a different name for the number of random samples from the posterior, such as `n_simulations`, `n_samples_from_posterior`, `n_random_samples`, or `n_draws`. [PyMC uses `draws` for this variable name.](https://pymc-devs.github.io/pymc3/api/inference.html#module-pymc3.sampling) @dswah what do you think?

dswah commented:
@cbrummitt mmm good call. i like shorter names, so im happiest with `n_draws` :)
What do you think?

cbrummitt commented:
How should we enable the user to control the `param_grids` of the gridsearch? 🤔  I see two options:

1. We use the default `param_grids={'lam': np.logspace(-3, 3, 11)}` always. The user cannot control it.
2. We augment gridsearch to handle distributions, like [sklearn.model_selection.RandomizedSearchCV](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html), so that a user can give to the `sample` method a dictionary mapping variable names to distributions.

The problem with naively enabling the user to provide an input such as `param_grids={'lam': np.logspace(-1, 1, 10)}` is that I think every bootstrap sample would have the same result (unless there's enough randomness in the fitting process?). I suspect that we want each of the `n_bootstrap` many `gridsearch`s to try a _different_ array of `lam` values.

dswah commented:
ahh crap, i see what you mean about trying *different* `lam` vectors...

dswah commented:
nice call

cbrummitt commented:
Just pushed my work that began with your great code snippet. I ended up splitting this calculation into a couple methods because it's long (and reading the book Clean Code has inspired me to create more single-purpose functions). Feel free to change that if you'd prefer to see the whole algorithm in one method.

Some new features to think carefully about:
- I enabled the user to choose what `quantity` to simulate: `coef`, `mu`, or `response`. Right now the API is to call `sample` and specify the `quantity` as a keyword argument.
- The user can optionally specify a different array of input data, `sample_at_X`, at which to draw samples (only relevant for `quantity` equal to `mu` or to `response`).

I made the default `n_bootstraps=1` because it's fast, doesn't create tons of progress bars, and may be all the user really wants, at least initially.

We may want to consider a `verbose` option to show progress on the bootstrap sampling loop in a way different from simply showing progress bars. Right now, if `n_bootstraps=20` (say), then 20 progress bars are shown sequentially, but there's no progress bar over the progress bars. Perhaps we could write messages of different levels of specificity, like the `verbose` parameter in `sklearn.model_selection.GridSearchCV`. (A `verbose` option in `gridsearch` is another thing that would be nice to have.)

I'm finished working on this for now, so feel free to work on it next.

cbrummitt commented on pygam/pygam.py:
```diff
@@ -1781,6 +1781,231 @@ def gridsearch(self, X, y, return_scores=False, keep_best=True,
+        # mapping bootstrap indices to draw indices and use the `size`
+        # parameter of `np.random.multivariate_normal` to sample the draws
+        # needed from that bootstrap sample all at once.
+        bootstrap_index_to_draw_indices = defaultdict(list)
```
The naive way to draw samples here would be:
```
coef_draws = np.empty((n_draws, len(self.coef_)))
for i, index_bootstrap in enumerate(random_indices_of_bootstraps):
    coef_draws[i] = np.random.multivariate_normal(
        coef_bootstraps[index_bootstrap],
        cov_bootstraps[index_bootstrap])
```

This fancy footwork using the dictionary `bootstrap_index_to_draw_indices` speeds things up by calling `np.random.multivariate_normal` only once for each bootstrap sample that gets used and by using the  `size`  parameter of `np.random.multivariate_normal` to draw all the samples it will ever need from that particular mean and covariance. Then the `[[draw_indices]]` puts those samples into the right rows. 🎉  💥

dswah commented:
hell yeah
i'll take a look at this tonight

cbrummitt commented:
Two ideas I've been thinking about:

- The `lam` values computed in `_bootstrap_samples_of_smoothing` are the computationally expensive part. Might we want to save those `lam` values in an attribute (say, `self.bootstrap_lam`) in case you want to efficiently sample some more in a way that reuses the bootstrap `lam` values, or in case you want to analyze the bootstrap `lam` values later? It could be like how `sklearn.model_selection.GridSearchCV` saves results in the `cv_results_` attribute. Then we'd have to think about how to let the user choose whether to append, reuse, or overwrite `self.bootstrap_lam` when calling the `sample` method.
- Right now the `gridsearch` uses the same sequence of values, `{'lam': np.logspace(-3, 3, 11)}`, which means the bootstrap sampling is pointless (unless the penalized fits are stochastic; are they?). We could accept a large sequence such as `{'lam': np.logspace(-3, 3, 100)}` and then partition it into `n_bootstrap` many subsets sampled uniformly without replacement. Then we'd have to think about how to handle other distributions such as `param_grid={'lam': np.logspace(-3, 3, 100), 'alpha': np.logspace(-3, 3, 30)}`.

cbrummitt commented:
Some TODOs:

- [ ] Fix the positive definite warnings using something like `cov = self.statistics_['cov'] + np.eye(n_features) * np.sqrt(EPS)` as described [in this comment above](https://github.com/dswah/pyGAM/pull/117#issuecomment-319220613)
- [ ] If we keep the `quantity='response'` API, then we should probably raise a `ValueError` if `quantity` is not in `{'response', 'mu', 'coef'}`. Right now, if you `sample(quantity='coefficient')` you'll get the response returned because it triggers the `else` statement.
- [ ] Add an example in the README
- [ ] Write some tests for shape and possibly for small examples with a random seed

cbrummitt commented:
@dswah Sorry for the hiatus. Is this PR still of interest? If so, thoughts on the `quantity='response'` API?

dswah commented:
@cbrummitt hey man! yeah, sorry as well.
i haven't had much time to dedicate to this project since we last spoke.
i definitely still like this PR, but i got hung up with other life things :)

the `quantitiy='` API seems cool to me.
i might just change `response` to `y` but other than that it seems intuitive

cbrummitt commented:
@dswah Did you fix the positive definite warnings (TODO item 1 in the comment 3 comments above this one)? I don't seem to get them anymore after I merged into my branch the changes that you made in master in September.

cbrummitt commented:
Should we add a parameter for whether to reuse the coefficients and covariance of the already fitted model? I wrote this TODO in the code:
```
# TODO: By reusing the attributes coef_ and statistics_['cov'] we
# are assuming that the user wants to reuse those. Make a parameter
# to toggle between reusing them or not?
```

cbrummitt commented on README.md:
```diff
@@ -121,7 +134,7 @@ for i, ax in enumerate(axs):
 
     ax.plot(XX[:, i], pdep)
-    ax.plot(XX[:, i], confi, c='r', ls='--')
+    ax.plot(XX[:, i], confi[0], c='r', ls='--')
```
`confi` is a list containing a 1D numpy array. I had to get that array to get this line to work; otherwise I got `ValueError: x and y must be the same size`

dswah replied:
awesome!!

cbrummitt commented:
I added to the README two examples of simulating from the posterior.

dswah commented:
@cbrummitt hmm im thinking about the`reuse_fitted_model_once` question..
i think we should aim for simplicity and fewer options and assume that the user knows what they want. 

so i suggest the following:
- always resuse the current model. bootstrap samples are expensive. lets keep as many as possible. lets assume the user will not change their minds about which search objective they want.

dswah commented:
@cbrummitt 
im starting to work on adding something like RandomSearchCV in order to accelerate model searches and get us better bootstrap samples. should be easy.

cbrummitt commented:
Sounds good to me. Let's not implement that `reuse_model_once` boolean parameter. That's something that we discussed here but did not implement in the code.

Right now, the user must make sure to call sample with the same objective they used when they fit the model before calling sample:
```
gam.gridsearch(X, y, objective='UBRE')
gam.sample(X, y, objective='UBRE')
```
I added an explanation in the docstring of `sample` that recommends doing gridsearch before doing `sample`. And I removed the TODO comment that was about the `reuse_fitted_model_once` idea. 

Regarding your step of adding a small amount to the diagonal, I made that a function in order to reuse that calculation for the first bootstrap sample of the coefficients and their covariance in the beginning of the method `_bootstrap_samples_of_smoothing`:
```
coef_bootstraps = [self.coef_]
cov_bootstraps = [_make_positive_semi_definite(self.statistics_['cov'])]
```
I just put that new function at the top of this script:
```
def _make_positive_semi_definite(cov):
        """Return the given matrix with a small amount added to the diagonal
        to make it positive semi-definite."""
        return cov + np.eye(len(cov)) * np.sqrt(EPS)
```
Should that function go in `utils.py` instead? If so, we'd have to define `EPS` in `utils.py`. I'll let you decide whether to keep this adding `EPS` to the diagonal as a separate function, how to document it, and where to put it. I'm happy to make changes you suggest or to merge your changes into this branch.

dswah commented on README.md:
```diff
@@ -151,6 +164,28 @@ Pseudo-R^2
+
+for i, ax in enumerate(axs):
+    pdep = gam.partial_dependence(XX, feature=i+1)
+    for sample_vector in samples:
```
@cbrummitt im just seeing this logistic regression chart for the first time. very cool graphic.
can you explain it to me? how come all the charts look so similar?

also, can you include a snippet that generated these samples?

cbrummitt replied:
Good catch. I wasn't thinking clearly enough. You're totally right: the vertical positions are exactly the same; only the horizontal scale is different in the three panels. I'd have to do a variant of `partial_dependence` that draws random samples, which doesn't seem worth it. I just removed this addition to the README. Just one illustration of `sample` (namely, `pygam_mcycle_data_linear_sample_from_posterior.png`) should be enough

dswah commented on pygam/distributions.py:
```diff
@@ -176,13 +200,35 @@ def deviance(self, y, mu, scaled=True):
     def __init__(self, levels=1):
         """
-        creates an instance of the NormalDist class
+        creates an instance of the Binomial class
```
:)

dswah commented on pygam/pygam.py:
```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+            `'mu'`, then samples are drawn at the values of `X` specified in
+            `sample_at_X`.
+
+        sample_at_X : array of shape (n_samples_to_simulate, m_features) or
```
👍

dswah commented on pygam/pygam.py:
```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+        None, default: None
+            Input data at which to draw new samples.
+
+            Only applies for `quantity == 'y'`. If `None`, then `sample_at_X`
```
also applies if `quantity == 'mu'` right?

cbrummitt replied:
Yep. I'll change that.

cbrummitt replied:
Fixed

dswah commented on pygam/pygam.py:
```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+
+        Returns
+        -------
+        draws : 2D array of length n_draws
```
👍

dswah commented on pygam/pygam.py:
```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+
+        for _ in range(n_bootstraps - 1):  # Wood pg. 198 step 2
+            # generate response data from fitted model (Wood pg. 198 step 3)
+            y_bootstrap = self.distribution.sample(mu)
```
hmm. actually im wondering about a technicality:
- wood says to sample errors/deviates from the fitted distribution
- add those errors to the original responses
- these are now our bootstraps

it seems the approach in the code is slightly different.
- we are essentially sampling deviates and adding them to mu instead of y

i think we will probably get more variation if we add deviates to the original responses?
what do you think?

dswah replied:
(reading pg 199, sec 4.9.3, step 3)

dswah replied:
you know what, never mind.

i think what you have is correct

dswah replied:
```
Based on the fitted model from step 1, simulate a parametric bootstrap response vector y[k]. That is, simulate random deviates with the appropriate response distribution, and mean given by the fitted values from step 1.
```

sounds like this means to sample from the original model. ie the way you have it.

dswah commented on pygam/pygam.py:
```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+        """Simulate coefficients using bootstrap samples."""
+        # Sample indices uniformly from {0, ..., n_bootstraps - 1}
+        # (Wood pg. 199 step 6)
+        random_bootstrap_indices = np.random.choice(
```
👍

dswah commented on pygam/pygam.py:
```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
+
+        coef_draws = np.empty((n_draws, len(self.coef_)))
+
+        for bootstrap, draw_indices in bootstrap_index_to_draw_indices.items():
```
cool. i finally get it :)

dswah commented on pygam/tests/test_GAM_methods.py:
```diff
@@ -327,4 +328,85 @@ def test_get_params_hidden():
     assert('_lam' in list(params.keys()))
 
+
+class TestSamplingFromPosterior(object):
```
i cant believe you wrote tests

dswah commented on pygam/tests/test_GAM_methods.py:
```diff
@@ -327,4 +328,85 @@ def test_get_params_hidden():
+
+class TestSamplingFromPosterior(object):
+
+    def test_drawing_samples_from_unfitted_model(self, mcycle, mcycle_gam):
```
good test name :)

dswah commented on pygam/tests/test_GAM_methods.py:
```diff
@@ -327,4 +328,85 @@ def test_get_params_hidden():
+    def test_sample_quantity(self, mcycle, mcycle_gam):
+        X, y = mcycle
+        for quantity in ['coefficients', 'response']:
+            with pytest.raises(ValueError):
```
cool. i've never seen this pattern before.
does this tell pytest to expect a ValueError?

cbrummitt replied:
Yep! Here's the [documentation](https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions). An example from there:
```
import pytest

def test_zero_division():
    with pytest.raises(ZeroDivisionError):
        1 / 0
```

dswah commented:
@cbrummitt man. this is really fucking awesome.

my only concern is that comment about the bootstrapped responses.
let me know what you think.

dswah commented:
aight cool, whats missing?

cbrummitt commented:
I think this PR is ready to go. `sample` is a bit trivial now with `gridsearch` reusing the same grid of parameters every time, as explained in this note in the docstring of `sample`:
```
NOTE: For now, the grid of `lam` values is the default of `gridsearch`.
Until randomized grid search is implemented, it is not worth setting
`n_bootstraps` to a value greater than one because the smoothing
parameters will be identical in each bootstrap sample.
```
Once your addition of randomized hyperparameter search gets merged, we should make that the default strategy used in the `sample` method.

Thanks for reviewing the PR!
